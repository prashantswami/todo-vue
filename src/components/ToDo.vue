<template>
    <div> 
        <StyledButton @click="openModal"> 
            Show Modal
        </StyledButton> 

        <h1>
            To do List
        </h1>
        <input type="text" name="add-task" class="form-control add-task" 
        v-model="newTask.name" 
        placeholder="Please add new task"
        @keyup.enter="addTask()">
        <div class="list">
            <div class="list-item" v-for="(task) in incompleteTasks" :key="task.taskId" @click="completeTask(task)">
                <span :class="{'completed' : task.isCompleted }">{{ task.name }}</span>
                <span class="remove" @click="removeTask(task.taskId)">&times;</span>
            </div>
        </div>
        <h4>Completed Tasks</h4>
        <div class="list">
            <div class="list-item" v-for="(task) in completedTasks" :key="task.taskId" @click="completeTask(task)">
                <span :class="{'completed' : task.isCompleted }">{{ task.name }}</span>
                <span class="remove" @click="removeTask(task.taskId)">&times;</span>
            </div>
        </div>


        <CommonModal v-model="showModal" > 
            <template v-slot:title>
                Hello
            </template>

            <template v-slot:body>
                body
            </template>

            <template v-slot:footer>
                <div class="footer">footer</div>
            </template>
        </CommonModal>
    </div>
</template>
<script>
import {mapState, mapActions, mapGetters} from 'vuex';
import StyledButton from './StyledButton.vue';
import CommonModal from './CommonModal.vue';

export default {
    name: "to-do",
    computed: {
        ...mapGetters({
            completedTasks: "task/completeTasks",
            incompleteTasks: "task/incompleteTasks"
        }),
        ...mapState({
            tasks: state => state.task.tasks
        })
    },
    components: {
        StyledButton,
        CommonModal
    },
    data() {
        return {
            newTask: {
                name: "",
                isCompleted: false,
            },
            latestTaskId: 0,
            showModal: false,
        };
    },
    mounted() {
        this.$fetchTasks();
    },
    methods: {
        ...mapActions({
            $fetchTasks: "task/fetchTasks",
            $addTask: "task/addTask",
            $removeTask: "task/removeTask",
            $clearAllTasks: "task/clearAllTasks",
            $updateTask: "task/updateTask",
        }),
        addTask() {
            if (this.newTask.name === "") {
                return;
            }
            this.latestTaskId = Math.random();
            this.$addTask({ ...this.newTask, taskId: this.latestTaskId });
            this.newTask.name = "";
        },
        removeTask(taskId) {
            this.$removeTask(taskId);
        },
        completeTask(task) {
            task.isCompleted = !task.isCompleted;
            this.$updateTask(task);
        },
        openModal() {
            this.showModal = true;
        }
    }
}
</script>

<style lang="scss" scoped>
.list, input {
    font-size: 18px;
}
.add-task {
    margin-bottom: 15px;
}
.form-control {
    width: 100%;
    padding: 15px;
}

.list {
    border-bottom: 1px solid rgb(211, 211, 211);
}

.list-item {
    text-align: left;
    padding: 15px 15px;
    border-left: 1px solid rgb(211, 211, 211);;
    border-right: 1px solid rgb(211, 211, 211);;
    border-top: 1px solid rgb(211, 211, 211);;
}
.list-item:hover {
    background-color: rgb(229, 247, 253);
}
.remove {
    float: right;
    cursor: pointer;
}
.completed {
    text-decoration: line-through;
}

.footer {
    padding: 15px;
}
</style>