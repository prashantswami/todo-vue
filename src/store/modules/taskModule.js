import TaskRestService from '../../rest/task.rest'

export const taskModule = {
    namespaced: true,
    state: {
        tasks: []
    },
    getters: {
        completeTasks(state) {
            return state.tasks.filter(task => task.isCompleted);
        },

        incompleteTasks(state) {
            return state.tasks.filter(task => !task.isCompleted);
        }
    },
    mutations: {
        FETCH_TASKS(state, payload) {
            state.tasks = payload;
        },

        ADD_TASK(state, payload) {
            state.tasks.push(payload)
        },

        REMOVE_TASK(state, taskId) {
            const foundTask = state.tasks.findIndex(t => t.taskId === taskId);
            if (foundTask !== -1) {
                state.tasks.splice(foundTask, 1);
            }
        },

        UPDATE_TASK(state, payload) {
            const foundTaskIndex = state.tasks.findIndex(t => t.taskId === payload.taskId);
            if (foundTaskIndex !== -1) {
                state.tasks.splice(foundTaskIndex, 1, { ...payload });
            }
        },

        CLEAR_ALL_TASKS(state) {
            state.tasks = [];
        }
    },
    actions: {
        fetchTasks(context) {
            const tasks = TaskRestService.get() ?? [];
            context.commit("FETCH_TASKS", tasks);
        },
        addTask(context, task) {
            context.commit("ADD_TASK", task);
            TaskRestService.update(context.state.tasks);
        },
        removeTask(context, taskId) {
            context.commit("REMOVE_TASK", taskId);
            TaskRestService.update(context.state.tasks);
        },
        updateTask(context, task) {
            context.commit("UPDATE_TASK", task);
            TaskRestService.update(context.state.tasks);
        },
        clearAllTasks(context) {
            context.commit("CLEAR_ALL_TASKS");
            TaskRestService.update(context.state.tasks);
        }
    },
}