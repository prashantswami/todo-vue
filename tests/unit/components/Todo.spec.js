import { shallowMount, createLocalVue } from '@vue/test-utils'
import ToDo from '@/components/ToDo.vue'
import Vuex from 'vuex';
import { taskModule } from '@/store/modules/taskModule'

const localVue = createLocalVue();

localVue.use(Vuex);

const data = {
    tasks: [
        {
            name: 'task 1',
            isCompleted: false,
            taskId: 5,
        }
    ]
}

const taskModule_ = {
    namespaced: true,
    state: {
        tasks: data.tasks,
    },
    getters: taskModule.getters,
    actions: {
        fetchTasks: jest.fn(),
        addTask: jest.fn(),
        removeTask: jest.fn(),
        updateTask: jest.fn(),
        clearAllTasks: jest.fn()
    },
}

const init = () => {
    const wrapper = shallowMount(ToDo, {
        localVue,
        store: new Vuex.Store({
            modules: {
                task: taskModule_
            }
        }),
    })

    return wrapper;
}

describe('ToDo.vue', () => {
    let wrapper;
    let that;

    const mountComp = () => {
        wrapper = init();
        that = wrapper.vm;
    }

    beforeEach(() => {
       
    })

    it('renders component', () => {
        mountComp();
        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.isVueInstance).toBeTruthy();
        expect(wrapper.vm.incompleteTasks).toMatchObject(data.tasks);
    })

    it('is able to add new task', ()=>{
        mountComp();
        that.newTask.name = "another task";
        that.addTask();

        expect(taskModule_.actions.addTask).toHaveBeenCalledTimes(1);
        expect(that.newTask.name).toBe('');
    })

    it('is able to remove that task', () => {
        mountComp();
        that.removeTask(5);
        expect(taskModule_.actions.removeTask).toHaveBeenCalledTimes(1);
    })

    it('is able to complete the task', () => {
        mountComp();
        that.completeTask(data.tasks[0]);
        expect(taskModule_.actions.updateTask).toBeCalledTimes(1);
        expect(taskModule_.actions.updateTask).toHaveBeenCalledWith(expect.any(Object), {
            ...data.tasks[0],
            isCompleted: true
        });
    })
})
