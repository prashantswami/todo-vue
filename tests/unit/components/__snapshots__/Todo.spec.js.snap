// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo.vue renders component 1`] = `
<div>
  <h1>To Do List 1</h1> <input type="text" name="add-task" placeholder="Please add new task" class="form-control add-task">
  <div class="list">
    <div class="list-item"><span class="">task 1</span> <span class="remove">Ã—</span></div>
  </div>
  <h4>Completed Tasks</h4>
  <div class="list"></div>
</div>
`;
